File: C#/WindowsPhone/LGame-XNA-lib/Loon.Core.Graphics.Component/LButton.cs
using Loon.Core.Graphics.Opengl;
using Loon.Utils;
using Loon.Core.Input;

namespace Loon.Core.Graphics.Component {
	
	public class LButton : LComponent {
	
		private string text = null;
	
		private bool over, pressed, exception;
	
		private int pressedTime, offsetLeft, offsetTop, type;
	
		private LFont font = LFont.GetDefaultFont();
	
		private LColor fontColor = LColor.white;
	
		public LButton(string fileName):this(fileName, null, 0, 0) {
			
		}
	
		public LButton(string fileName, string text, int x, int y):this(new LTexture(fileName), text, x, y) {
			
		}
	
		public LButton(LTexture img, string text, int x, int y):this(img, text, img.GetWidth(), img.GetHeight(), x, y) {
			
		}
	
		public LButton(string fileName, int row, int col):this(fileName, null, row, col, 0, 0) {
			
		}
	
		public LButton(string fileName, string text, int row, int col, int x, int y):this(LTextures.LoadTexture(fileName), text, row, col, x, y) {
		
		}
	
		public LButton(LTexture img, string text, int row, int col, int x, int y):this(TextureUtils.GetSplitTextures(img, row, col), text, row, col, x, y) {
			
		}
	
		public LButton(LTexture[] img, string text, int row, int col, int x, int y):base(x, y, row, col) {
			this.SetImages(img);
			this.text = text;
		}
	
		public LButton(string text, int x, int y, int w, int h):base(x, y, w, h) {
			this.text = text;
		}

        public virtual void SetImages(params LTexture[] images)
        {
			LTexture[] buttons = new LTexture[4];
			if (images != null) {
				int size = images.Length;
				this.type = size;
				switch (size) {
				case 1:
					buttons[0] = images[0];
					buttons[1] = images[0];
					buttons[2] = images[0];
					buttons[3] = images[0];
					break;
				case 2:
					buttons[0] = images[0];
					buttons[1] = images[1];
					buttons[2] = images[0];
					buttons[3] = images[0];
					break;
				case 3:
					buttons[0] = images[0];
					buttons[1] = images[1];
					buttons[2] = images[2];
					buttons[3] = images[0];
					break;
				case 4:
					buttons = images;
					break;
				default:
					exception = true;
					break;
				}
			}
			if (!exception) {
				this.SetImageUI(buttons, true);
			}
	
		}
	
		public override void CreateUI(GLEx g, int x, int y, LComponent component,
				LTexture[] buttonImage) {
			LButton button = (LButton) component;
			if (buttonImage != null) {
				if (!button.IsEnabled()) {
					g.DrawTexture(buttonImage[3], x, y);
				} else if (button.IsTouchPressed()) {
					g.DrawTexture(buttonImage[2], x, y);
				} else if (button.IsTouchOver()) {
					g.DrawTexture(buttonImage[1], x, y);
				} else {
					if (type == 1) {
						g.DrawTexture(buttonImage[0], x, y, LColor.gray);
					} else {
						g.DrawTexture(buttonImage[0], x, y);
					}
				}
			}
			if (text != null) {
				LFont old = g.GetFont();
				g.SetFont(font);
				g.SetColor(fontColor);
				g.DrawString(
						text,
						x + button.GetOffsetLeft()
								+ (button.GetWidth() - font.StringWidth(text)) / 2,
						y + button.GetOffsetTop()
								+ (button.GetHeight() - font.GetLineHeight()) / 2
								+ font.GetLineHeight());
				g.SetFont(old);
				g.ResetColor();
			}
		}

        public override void Update(long timer)
        {
			if (this.pressedTime > 0 && --this.pressedTime <= 0) {
				this.pressed = false;
			}
		}
	
		public bool IsTouchOver() {
			return this.over;
		}
	
		public bool IsTouchPressed() {
			return this.pressed;
		}
	
		public string GetText() {
			return this.text;
		}
	
		public void SetText(string st) {
			this.text = st;
		}
	
		protected internal override void ProcessTouchDragged() {
			if (this.input.GetTouchPressed() == Touch.TOUCH_MOVE) {
				this.over = this.pressed = this.Intersects(this.input.GetTouchX(),
						this.input.GetTouchY());
			}
		}
	
		public void DoClick() {
			if (Click != null) {
                Click.DoClick(this);
			}
		}
	
		public void DownClick() {
			if (Click != null) {
				Click.DownClick(this, input.GetTouchX(), input.GetTouchY());
			}
		}
	
		public void UpClick() {
			if (Click != null) {
				Click.UpClick(this, input.GetTouchX(), input.GetTouchY());
			}
		}

        protected internal override void ProcessTouchClicked()
        {
			if (this.input.GetTouchReleased() == Touch.TOUCH_UP) {
				this.DoClick();
			}
		}

        protected internal override void ProcessTouchPressed()
        {
			if (this.input.GetTouchPressed() == Touch.TOUCH_DOWN) {
				this.DownClick();
				this.pressed = true;
			}
		}

        protected internal override void ProcessTouchReleased()
        {
			if (this.input.GetTouchReleased() == Touch.TOUCH_UP) {
				this.UpClick();
				this.pressed = false;
			}
		}

        protected internal override void ProcessTouchEntered()
        {
			this.over = true;
		}

        protected internal override void ProcessTouchExited()
        {
			this.over = this.pressed = false;
		}

        protected internal override void ProcessKeyPressed()
        {
			if (this.IsSelected() && this.input.GetKeyPressed() == Key.ENTER) {
				this.pressedTime = 5;
				this.pressed = true;
				this.DoClick();
			}
		}

        protected internal override void ProcessKeyReleased()
        {
			if (this.IsSelected() && this.input.GetKeyReleased() == Key.ENTER) {
				this.pressed = false;
			}
		}

        public virtual bool IsException()
        {
			return exception;
		}

        public override string GetUIName()
        {
			return "Button";
		}

        public virtual LFont GetFont()
        {
			return font;
		}

        public virtual void SetFont(LFont font)
        {
			this.font = font;
		}

        public virtual LColor GetFontColor()
        {
			return fontColor;
		}

        public virtual void SetFontColor(LColor fontColor)
        {
			this.fontColor = fontColor;
		}

        public virtual int GetOffsetLeft()
        {
			return offsetLeft;
		}

        public virtual void SetOffsetLeft(int offsetLeft)
        {
			this.offsetLeft = offsetLeft;
		}

        public virtual int GetOffsetTop()
        {
			return offsetTop;
		}

        public virtual void SetOffsetTop(int offsetTop)
        {
			this.offsetTop = offsetTop;
		}
	
	}
}

File: Java/Android/OpenGLES/source/core/src/loon/core/graphics/component/LButton.java
package loon.core.graphics.component;

import loon.core.graphics.LColor;
import loon.core.graphics.LComponent;
import loon.core.graphics.LFont;
import loon.core.graphics.opengl.GLEx;
import loon.core.graphics.opengl.LTexture;
import loon.core.graphics.opengl.LTextures;
import loon.core.graphics.opengl.TextureUtils;
import loon.core.input.LInputFactory.Key;
import loon.core.input.LInputFactory.Touch;


/**
 * 
 * Copyright 2008 - 2009
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.1
 */

public class LButton extends LComponent {

	private String text = null;

	private boolean over, pressed, exception;

	private int pressedTime, offsetLeft, offsetTop, type;

	private LFont font = LFont.getDefaultFont();

	private LColor fontColor = LColor.white;

	public LButton(String fileName) {
		this(fileName, null, 0, 0);
	}

	public LButton(String fileName, String text, int x, int y) {
		this(new LTexture(fileName), text, x, y);
	}

	public LButton(LTexture img, String text, int x, int y) {
		this(img, text, img.getWidth(), img.getHeight(), x, y);
	}

	public LButton(String fileName, int row, int col) {
		this(fileName, null, row, col, 0, 0);
	}

	public LButton(String fileName, String text, int row, int col, int x, int y) {
		this(LTextures.loadTexture(fileName), text, row, col, x, y);
	}

	public LButton(LTexture img, String text, int row, int col, int x, int y) {
		this(TextureUtils.getSplitTextures(img, row, col), text, row, col, x, y);
	}

	public LButton(LTexture[] img, String text, int row, int col, int x, int y) {
		super(x, y, row, col);
		this.setImages(img);
		this.text = text;
	}

	public LButton(String text, int x, int y, int w, int h) {
		super(x, y, w, h);
		this.text = text;
	}

	public void setImages(LTexture... images) {
		LTexture[] buttons = new LTexture[4];
		if (images != null) {
			int size = images.length;
			this.type = size;
			switch (size) {
			case 1:
				buttons[0] = images[0];
				buttons[1] = images[0];
				buttons[2] = images[0];
				buttons[3] = images[0];
				break;
			case 2:
				buttons[0] = images[0];
				buttons[1] = images[1];
				buttons[2] = images[0];
				buttons[3] = images[0];
				break;
			case 3:
				buttons[0] = images[0];
				buttons[1] = images[1];
				buttons[2] = images[2];
				buttons[3] = images[0];
				break;
			case 4:
				buttons = images;
				break;
			default:
				exception = true;
				break;
			}
		}
		if (!exception) {
			this.setImageUI(buttons, true);
		}

	}

	@Override
	public void createUI(GLEx g, int x, int y, LComponent component,
			LTexture[] buttonImage) {
		LButton button = (LButton) component;
		if (buttonImage != null) {
			if (!button.isEnabled()) {
				g.drawTexture(buttonImage[3], x, y);
			} else if (button.isTouchPressed()) {
				g.drawTexture(buttonImage[2], x, y);
			} else if (button.isTouchOver()) {
				g.drawTexture(buttonImage[1], x, y);
			} else {
				if (type == 1) {
					g.drawTexture(buttonImage[0], x, y, LColor.gray);
				} else {
					g.drawTexture(buttonImage[0], x, y);
				}
			}
		}
		if (text != null) {
			LFont old = g.getFont();
			g.setFont(font);
			g.setColor(fontColor);
			g.drawString(
					text,
					x + button.getOffsetLeft()
							+ (button.getWidth() - font.stringWidth(text)) / 2,
					y + button.getOffsetTop()
							+ (button.getHeight() - font.getLineHeight()) / 2
							+ font.getLineHeight());
			g.setFont(old);
			g.resetColor();
		}
	}

	@Override
	public void update(long timer) {
		if (this.pressedTime > 0 && --this.pressedTime <= 0) {
			this.pressed = false;
		}
	}

	public boolean isTouchOver() {
		return this.over;
	}

	public boolean isTouchPressed() {
		return this.pressed;
	}

	public String getText() {
		return this.text;
	}

	public void setText(String st) {
		this.text = st;
	}

	@Override
	protected void processTouchDragged() {
		if (this.input.getTouchPressed() == Touch.TOUCH_MOVE) {
			this.over = this.pressed = this.intersects(this.input.getTouchX(),
					this.input.getTouchY());
		}
	}

	/**
	 * 处理点击事件（请重载实现）
	 * 
	 */
	public void doClick() {
		if (Click != null) {
			Click.DoClick(this);
		}
	}

	public void downClick() {
		if (Click != null) {
			Click.DownClick(this,input.getTouchX(), input.getTouchY());
		}
	}

	public void upClick() {
		if (Click != null) {
			Click.UpClick(this,input.getTouchX(), input.getTouchY());
		}
	}

	@Override
	protected void processTouchClicked() {
		if (this.input.getTouchReleased() == Touch.TOUCH_UP) {
			this.doClick();
		}
	}

	@Override
	protected void processTouchPressed() {
		if (this.input.getTouchPressed() == Touch.TOUCH_DOWN) {
			this.downClick();
			this.pressed = true;
		}
	}

	@Override
	protected void processTouchReleased() {
		if (this.input.getTouchReleased() == Touch.TOUCH_UP) {
			this.upClick();
			this.pressed = false;
		}
	}

	@Override
	protected void processTouchEntered() {
		this.over = true;
	}

	@Override
	protected void processTouchExited() {
		this.over = this.pressed = false;
	}

	@Override
	protected void processKeyPressed() {
		if (this.isSelected() && this.input.getKeyPressed() == Key.ENTER) {
			this.pressedTime = 5;
			this.pressed = true;
			this.doClick();
		}
	}

	@Override
	protected void processKeyReleased() {
		if (this.isSelected() && this.input.getKeyReleased() == Key.ENTER) {
			this.pressed = false;
		}
	}

	public boolean isException() {
		return exception;
	}

	@Override
	public String getUIName() {
		return "Button";
	}

	public LFont getFont() {
		return font;
	}

	public void setFont(LFont font) {
		this.font = font;
	}

	public LColor getFontColor() {
		return fontColor;
	}

	public void setFontColor(LColor fontColor) {
		this.fontColor = fontColor;
	}

	public int getOffsetLeft() {
		return offsetLeft;
	}

	public void setOffsetLeft(int offsetLeft) {
		this.offsetLeft = offsetLeft;
	}

	public int getOffsetTop() {
		return offsetTop;
	}

	public void setOffsetTop(int offsetTop) {
		this.offsetTop = offsetTop;
	}

}
